Feature/print property list

kennyweiss opened the pull request:
Added a new blt macro to print all cmake and blt properties for a given target, which should be a helpful diagnostic (e.g. for #184 )

Here is some sample output for a registered library that is also a target (gtest), a static library (example), a header only library (blt_header_only) and an executable (t_example_smoke):
```cmake
-- [gtest property] BINARY_DIR: <blt_root>/build-blt/blt/thirdparty_builtin/googletest-master-2018-02-01/googletest
-- [gtest property] BUILD_WITH_INSTALL_RPATH: OFF
-- [gtest property] COMPILE_FLAGS:      -Wall -Wextra  -Wall -Wshadow -Werror -DGTEST_HAS_PTHREAD=1 -fexceptions -Wextra -Wno-unused-parameter -Wno-missing-field-initializers
-- [gtest property] CXX_EXTENSIONS: OFF
-- [gtest property] CXX_STANDARD: 11
-- [gtest property] CXX_STANDARD_REQUIRED: ON
-- [gtest property] Fortran_MODULE_DIRECTORY: <blt_root>/build-blt/lib/fortran
-- [gtest property] IMPORTED: FALSE
-- [gtest property] INCLUDE_DIRECTORIES: <blt_root>/thirdparty_builtin/googletest-master-2018-02-01/googletest/include;<blt_root>/thirdparty_builtin/googletest-master-2018-02-01/googletest
-- [gtest property] INSTALL_RPATH:
-- [gtest property] INSTALL_RPATH_USE_LINK_PATH: OFF
-- [gtest property] INTERFACE_INCLUDE_DIRECTORIES: <blt_root>/thirdparty_builtin/googletest-master-2018-02-01/googletest/include
-- [gtest property] INTERFACE_LINK_LIBRARIES: -pthread
-- [gtest property] INTERFACE_SYSTEM_INCLUDE_DIRECTORIES: <blt_root>/thirdparty_builtin/googletest-master-2018-02-01/googletest/include
-- [gtest property] LINK_LIBRARIES: -pthread
-- [gtest property] NAME: gtest
-- [gtest property] POSITION_INDEPENDENT_CODE: TRUE
-- [gtest property] RUNTIME_OUTPUT_DIRECTORY: <blt_root>/build-blt/bin
-- [gtest property] SKIP_BUILD_RPATH: OFF
-- [gtest property] SOURCES: src/gtest-all.cc
-- [gtest property] SOURCE_DIR: <blt_root>/thirdparty_builtin/googletest-master-2018-02-01/googletest
-- [gtest property] TYPE: STATIC_LIBRARY
-- [gtest property] BLT_GTEST_COMPILE_FLAGS:
-- [gtest property] BLT_GTEST_DEFINES: -DGTEST_HAS_DEATH_TEST=0
-- [gtest property] BLT_GTEST_INCLUDES: <blt_root>/thirdparty_builtin/googletest-master-2018-02-01/googletest/include
-- [gtest property] BLT_GTEST_LIBRARIES: gtest_main;gtest;-pthread
-- [gtest property] BLT_GTEST_TREAT_INCLUDES_AS_SYSTEM: ON
-- [example property] BINARY_DIR: <blt_root>/build-blt
-- [example property] BUILD_WITH_INSTALL_RPATH: OFF
-- [example property] CXX_EXTENSIONS: OFF
-- [example property] CXX_STANDARD: 11
-- [example property] CXX_STANDARD_REQUIRED: ON
-- [example property] Fortran_MODULE_DIRECTORY: <blt_root>/build-blt/lib/fortran
-- [example property] IMPORTED: FALSE
-- [example property] INSTALL_RPATH:
-- [example property] INSTALL_RPATH_USE_LINK_PATH: OFF
-- [example property] NAME: example
-- [example property] POSITION_INDEPENDENT_CODE: TRUE
-- [example property] RUNTIME_OUTPUT_DIRECTORY: <blt_root>/build-blt/bin
-- [example property] SKIP_BUILD_RPATH: OFF
-- [example property] SOURCES: src/Example.cpp;src/Example.hpp
-- [example property] SOURCE_DIR: <blt_root>/tests/internal
-- [example property] TYPE: STATIC_LIBRARY
-- [t_example_smoke property] BINARY_DIR: <blt_root>/build-blt
-- [t_example_smoke property] BUILD_WITH_INSTALL_RPATH: OFF
-- [t_example_smoke property] COMPILE_DEFINITIONS: GTEST_HAS_DEATH_TEST=0
-- [t_example_smoke property] CXX_EXTENSIONS: OFF
-- [t_example_smoke property] CXX_STANDARD: 11
-- [t_example_smoke property] CXX_STANDARD_REQUIRED: ON
-- [t_example_smoke property] Fortran_MODULE_DIRECTORY: <blt_root>/build-blt/lib/fortran
-- [t_example_smoke property] IMPORTED: FALSE
-- [t_example_smoke property] INCLUDE_DIRECTORIES: <blt_root>/thirdparty_builtin/googletest-master-2018-02-01/googletest/include
-- [t_example_smoke property] INSTALL_RPATH:
-- [t_example_smoke property] INSTALL_RPATH_USE_LINK_PATH: OFF
-- [t_example_smoke property] INTERFACE_COMPILE_DEFINITIONS: GTEST_HAS_DEATH_TEST=0
-- [t_example_smoke property] INTERFACE_INCLUDE_DIRECTORIES: <blt_root>/thirdparty_builtin/googletest-master-2018-02-01/googletest/include
-- [t_example_smoke property] INTERFACE_LINK_LIBRARIES: example;gtest_main;gtest;-pthread
-- [t_example_smoke property] INTERFACE_SYSTEM_INCLUDE_DIRECTORIES: <blt_root>/thirdparty_builtin/googletest-master-2018-02-01/googletest/include
-- [t_example_smoke property] LINK_LIBRARIES: example;gtest_main;gtest;-pthread
-- [t_example_smoke property] NAME: t_example_smoke
-- [t_example_smoke property] POSITION_INDEPENDENT_CODE: TRUE
-- [t_example_smoke property] RUNTIME_OUTPUT_DIRECTORY: <blt_root>/build-blt/bin
-- [t_example_smoke property] SKIP_BUILD_RPATH: OFF
-- [t_example_smoke property] SOURCES: src/t_example_smoke.cpp
-- [t_example_smoke property] SOURCE_DIR: <blt_root>/tests/internal
-- [t_example_smoke property] TYPE: EXECUTABLE
-- [blt_header_only property] INTERFACE_SOURCES: $<BUILD_INTERFACE:<blt_root>/tests/internal/src/HeaderOnly.hpp>
-- [blt_header_only property] NAME: blt_header_only
-- [blt_header_only property] TYPE: INTERFACE_LIBRARY
```

white238 commented on cmake/BLTMacros.cmake:
```diff
@@ -979,3 +979,86 @@ macro(blt_combine_static_libraries)
+        message(FATAL_ERROR "TARGET is a required parameter for the blt_print_target_properties macro")
+    endif()
+
+    if(TARGET ${arg_TARGET})
```
You should error out if not a cmake or blt registered target

kennyweiss replied:
I think it should be ok here since it is a diagnostic.

On the flip side -- Is there a way to check if something is a registered target?  I guess it has ``BLT_FOO_INCLUDES`` defined?

white238 replied:
True... I am fine with it not erroring out but can we add a message at the start and/or end that indicates it is printing information.  That way when its empty they know something has gone wrong.

kennyweiss replied:
I'll add a comment to indicate whether the target is a cmake target, a registered target,  neither or both.

white238 replied:
Fantastic.

kennyweiss replied:
Done.

Here is the output for 'not-a-target':
```cmake
-- [blt_print_target_properties] Invalid argument 'not-a-target'. This macro applies only to valid cmake targets or blt_registered targets.
```

kennyweiss commented on cmake/BLTMacros.cmake:
```diff
@@ -979,3 +979,86 @@ macro(blt_combine_static_libraries)
+        execute_process(COMMAND cmake --help-property-list OUTPUT_VARIABLE _property_list)
+        string(REGEX REPLACE ";" "\\\\;" _property_list "${_property_list}")
+        string(REGEX REPLACE "\n" ";" _property_list "${_property_list}")
+        list(FILTER _property_list EXCLUDE REGEX "^LOCATION$|^LOCATION_|_LOCATION$")
```
Looks like ``list(FILTER )`` is only available in ``cmake-3.6+`` and our travis is ``cmake-3.3.1``.
I'll find an alternative...

white238 commented:
After some thought, BLT_*_INCLUDES is not enough. For example, Fortran libraries do not have an include directory.

kennyweiss commented:
@white238 -- I added a new output variable to ``blt_register_library`` -- ``BLT_FOO_IS_REGISTERED_LIBRARY`` which is now being used in ``blt_print_target_properties``.

I also implemented a replacement for ``list(FILTER ..)`` as a ``BLTPrivateMacro`` so that we don't need to require a newer version of cmake.
